pkg/test/framework/suite.go

```diff
@@ -0,0 +1,254 @@
+)
+
+var (
+	rt   *runtime.Instance
+	rtMu sync.Mutex
+)
+
+// SetupFn is a function used for performing suite-level setup actions.
+type SetupFn func(ctx SuiteContext) error
+
+// mRunFn abstracts testing.M.run, so that the framework itself can be tested.
+type mRunFn func() int
+
+// Suite allows the test author to specify suite-related metadata and do setup in a fluent-style, before commencing execution.
+type Suite struct {
+	testID string
+	mRun   mRunFn
+	osExit func(int)
+	labels []label.Instance
+
+	setupFns []SetupFn
+}
+
+// NewSuite returns a new suite instance.
+func NewSuite(testID string, m *testing.M) *Suite {
+	return newSuite(testID, m.Run, os.Exit)
+}
+
+func newSuite(testID string, fn mRunFn, osExit func(int)) *Suite {
+	s := &Suite{
+		testID: testID,
+		mRun:   fn,
+		osExit: osExit,
+	}
+
+	return s
+}
+
+// Label all the tests in suite with the given labels
+func (s *Suite) Label(labels ...label.Instance) *Suite {
+	s.labels = append(s.labels, labels...)
+	return s
+}
+
+// RequireEnvironment ensures that the current environment matches what the suite expects. Otherwise it
+// stops test execution. This also applies the appropriate label to the suite implicitly.
+func (s *Suite) RequireEnvironment(name environment.Name) *Suite {
+	setupFn := func(ctx SuiteContext) error {
+		if name != ctx.Environment().EnvironmentName() {
+			scopes.Framework.Infof("Skipping suite %q: Required environment (%v) does not match current: %v",
+				ctx.Settings().TestID, name, ctx.Environment().EnvironmentName())
+			s.osExit(0)
+		}
+		return nil
+	}
+
+	// Prepend the function, so that it runs as the first thing.
+	fns := []SetupFn{setupFn}
+	fns = append(fns, s.setupFns...)
+	s.setupFns = fns
+	return s
+}
+
+// Setup runs enqueues the given setup function to run before test execution.
+func (s *Suite) Setup(fn SetupFn) *Suite {
+	s.setupFns = append(s.setupFns, fn)
+	return s
+}
+
+func (s *Suite) runSetupFn(fn SetupFn, ctx SuiteContext) (err error) {
+	defer func() {
+		// Dump if the setup function fails
+		if err != nil && ctx.Settings().CIMode {
+			rt.Dump()
+		}
+	}()
+	err = fn(ctx)
+	return
+}
+
+// EnvSetup runs the given setup function conditionally, based on the current environment.
+func (s *Suite) EnvSetup(e environment.Name, fn SetupFn) *Suite {
+	s.Setup(func(ctx SuiteContext) error {
+		if ctx.Environment().EnvironmentName() != e {
+			return nil
+		}
+		return fn(ctx)
+	})
+
+	return s
+}
+
+// Run the suite. This method calls os.Exit and does not return.
+func (s *Suite) Run() {
+	s.osExit(s.run())
+}
+
+func (s *Suite) run() (errLevel int) {
+	if err := initRuntime(s.testID, s.labels); err != nil {
+		scopes.Framework.Errorf("Error during test framework init: %v", err)
+		return exitCodeInitError
+	}
+
+	ctx := rt.SuiteContext()
+
+	// Before starting, check whether the current set of labels & label selectors will ever allow us to run tests.
+	// if not, simply exit now.
+	if ctx.Settings().Selector.Excludes(s.labels) {
```

Comment:
Would `!selector.Selects(s.labels)` work here as well?

Actually `Selects` might even be more correct. If there are positive labels on the suite and you don't run with that label, I would expect that you shouldn't run the suite.

Fundamentally, I don't think Suites should be any different from Tests WRT labels.  Suites are just composite tests.

Reply:
It should not be a strict match. The filter is a bunch of predicates (+ve and -ve). If there is no filter, everything gets selected. This is how we can perform things like "everything except" with a negative predicate.

That is why Excludes is needed. The combination of suite + test labels can cause the selection to work. But there are cases where you can tell the selection will not work solely based on suite labels. If, at the suite level, it is guaranteed that the selection will not work, then you can skip running suite setup (which is costly). 

The alternative is to add more explicit logical construction with ors and ands and possibly prefix/regex matches on labels. That is more flexible. We can add them in the future.

Reply:
Would you ever have positive labels on the suite then? It doesn't seem like there would be a point in doing so, since they wouldn't be used in the current logic AFAICT.

Reply:
Yes. You can say "run this whole suite as Presubmit, or Postsubmit". Applying labels is purely additive.

Reply:
Can you point me to where a suite is positively selected? I must be missing it.

Reply:
The selection is happening for the combination of  Suite+Test labels, so there is no suite level selection logic anywhere. A single suite, by itself will not positively getting selected. The selection applies to tests only.

However, before applying the selector to each and every test, at the suite level, you can optimize by asking "even if I add more labels to the ones specified by the suite, will it ever select?" 

The example is, if I tag the whole suite as "presubmit", and I don't want to run presubmit (-presubmit), then I know no other test in the suite will get selected. I can check and exit immediately, without running the test setup (i.e. deployment).

Conversely, I can tag a suite with presubmit, and selector can have presubmit (+presubmit), this doesn't guarantee that all the tests will be selected, and some maybe filtered out. For example, if we had a hypothetical "flaky" label, then you can tag a suite as presubmit, a test as flaky and have a selector that says (+presubmit,-flaky) and not run the test.